File: hub-backend/src/main/java/org/panda_lang/hub/user/UserConfiguration.java
/*
 * Copyright (c) 2018-2019 Hub Team
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.panda_lang.hub.user;

import org.panda_lang.hub.authentication.OAuth2UserDetailsService;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.security.crypto.password.PasswordEncoder;

@Configuration
class UserConfiguration {

    @Bean
    UserDetailsService authenticationUserDetailsService(UserRepository userRepository) {
        return new UserDetailsService(userRepository);
    }

    @Bean
    OAuth2UserDetailsService defaultOAuth2UserService(UserRepository userRepository) {
        return new UserOAuth2DetailsService(userRepository);
    }

    @Bean
    UserFacade userFacade(PasswordEncoder passwordEncoder, UserRepository userRepository) {
        return new UserFacade(passwordEncoder, userRepository);
    }

}

File: hub-backend/src/main/kotlin/org/panda_lang/hub/user/UserConfiguration.kt
/*
 * Copyright (c) 2021 Hub Team of panda-lang organization
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.panda_lang.hub.user

import io.ktor.application.Application
import io.ktor.client.HttpClient
import io.ktor.routing.Routing
import org.litote.kmongo.coroutine.CoroutineDatabase
import org.panda_lang.hub.github.RemoteGitHubClient

fun Application.usersModule(httpClient: HttpClient, database: CoroutineDatabase): UserFacade {
    val gitHubClient = RemoteGitHubClient(httpClient)

    val userCollection = database.getCollection<User>()
    val userRepository = MongoUserRepository(userCollection)
    val userService = UserService(gitHubClient, userRepository)

    return UserFacade(userService)
}

fun installUserRouting(routing: Routing, userFacade: UserFacade) {
    routing.routes(userFacade)
}
